src/go-virtual-machine-main/virtual-machine/executor/instructions/store.ts

```diff
@@ -55,38 +57,82 @@ export class StoreArrayElementInstruction extends Instruction {
-      process.heap.copy(elemAddr, src)
-    }
-    //const array = new ArrayNode(process.heap, dst)
-    /*
-    if (this.index < 0 || this.index >= array.length()) {
-      throw new Error(
-        `Index out of range [${this.index}] with length ${array.length()}`,
-      )
-    }
-    */
+      const node = process.heap.get_value(dst)
+      if (node instanceof StructNode || node instanceof ArrayNode) {
+        let elemAddr = this.peek(process, node, this.index, 0)
+        if (elemAddr instanceof BaseNode) elemAddr = elemAddr.addr
+        process.heap.copy(elemAddr, src)
+      }
+      //const array = new ArrayNode(process.heap, dst)
+      /*
+      if (this.index < 0 || this.index >= array.length()) {
+        throw new Error(
+          `Index out of range [${this.index}] with length ${array.length()}`,
+        )
+      }
+      */
 
-    if (process.debug_mode) {
-      process.debugger.modified_buffer.add(dst)
+      if (process.debug_mode) {
+        process.debugger.modified_buffer.add(dst)
+      }
+    } else {
+      // if it is not the first field and we know that it is not to be popped from OS,
+      // we have to get rid of the debris empty array and then flip the other 2 nodes around
+      // Old OS: ..., correct array, value to store in next field, residual empty array
+      // Correct version: ..., value to store in next index, correct array
+      // Since we are popping the other 2 nodes anyway, just flip the variable names around
+      if (this.index > 0) {
+        process.context.popOS() // empty residual array from previous LoadVariableInstruction
+        const src = process.context.popOS()
+        let dst = process.context.popOS()
+        if (process.heap.get_value(dst) instanceof ReferenceNode) {
+          dst = (process.heap.get_value(dst) as ReferenceNode).get_child()
+        }
+        const node = process.heap.get_value(dst)
+        if (node instanceof StructNode || node instanceof ArrayNode) {
+          let elemAddr = this.peek(process, node, this.index, 0)
+          if (elemAddr instanceof BaseNode) elemAddr = elemAddr.addr
+          process.heap.copy(elemAddr, src)
+        }
+
+        if (process.debug_mode) {
+          process.debugger.modified_buffer.add(dst)
+        }
+        process.context.pushOS(dst)
+      } else {
+        let dst = process.context.popOS()
+        const src = process.context.popOS()
+        if (process.heap.get_value(dst) instanceof ReferenceNode) {
+          dst = (process.heap.get_value(dst) as ReferenceNode).get_child()
+        }
+        const struct = new StructNode(process.heap, dst)
+        let fieldAddr = this.peek(process, struct, this.index, 0)
+        if (fieldAddr instanceof BaseNode) fieldAddr = fieldAddr.addr
+        process.heap.copy(fieldAddr, src)
+
+        if (process.debug_mode) {
+          process.debugger.modified_buffer.add(dst)
+        }
+        process.context.pushOS(dst)
+      }
```

Comment:
First, I think this is too repetitive, secondly. Why does in the case index = 0, you're peeking the Struct while in the other case you're checking if it's an ```ArrayNode``` or ```StructNode```.

If there are no particular reason for this, I think that you can handle the code by just doing the pop residual array if index is > 0 and keep the remaining the same.
```suggestion
        if (this.index > 0) {
          process.context.popOS() // empty residual struct from previous LoadVariableInstruction
        }
        const src = process.context.popOS()
        let dst = process.context.popOS()
        if (process.heap.get_value(dst) instanceof ReferenceNode) {
          dst = (process.heap.get_value(dst) as ReferenceNode).get_child()
        }
        if (node instanceof StructNode || node instanceof ArrayNode) {
          let elemAddr = this.peek(process, node, this.index, 0)
          if (elemAddr instanceof BaseNode) elemAddr = elemAddr.addr
          process.heap.copy(elemAddr, src)
        }

        if (process.debug_mode) {
          process.debugger.modified_buffer.add(dst)
        }
        process.context.pushOS(dst)
      }
```

Reply:
Hello there,
The checking of whether it's an `ArrayNode` or `StructNode` is unintentional, it should have been to just create an array directly (embarrassingly I mixed up with structs, that is supposed to be an array). This is to handle the case you mentioned earlier where we create the struct at the same time as passing it as an argument to another function, so it does not have a variable name to store to, thus it also does not appear in the environment frame of the caller.

As for popping, we have to see which is the destination (copy data to) and which is the source (copy data from), since we are using an operand stack, so depending on whether there is a residual array/struct to work with, we pop them reversely - If the previous operation left it on the stack, that would be popped later and thus shall be the destination, and the one popped earlier shall be the source.
- If the previous operation did not leave it on the stack because it has a variable to store the whole array/struct to, that would be loaded later and thus popped earlier to become the destination, and the one popped later shall be the source.

Other than that yes, the code that follows are the same, so it has been simplified as of the latest commit.